File: hphp/runtime/ext/mailparse/ext_mailparse.php
<?hh
/* Create a MIME mail resource.
 * @return resource - Returns a handle that can be used to parse a message.
 */
<<__Native>>
function mailparse_msg_create(): resource;

/* Frees a MIME resource.
 * @param resource $mimemail - A valid MIME resource allocated by
 * mailparse_msg_create() or mailparse_msg_parse_file().
 * @return bool - Returns TRUE on success or FALSE on failure.
 */
<<__Native>>
function mailparse_msg_free(resource $mimemail): bool;

/* Parses a file. This is the optimal way of parsing a mail file that you have
 * on disk.
 * @param string $filename - Path to the file holding the message. The file is
 * opened and streamed through the parser.
 * @return mixed - Returns a MIME resource representing the structure, or
 * FALSE on error.
 */
<<__Native>>
function mailparse_msg_parse_file(string $filename): mixed;

/* Incrementally parse data into the supplied mime mail resource.  This
 * function allow you to stream portions of a file at a time, rather than read
 * and parse the whole thing.
 * @param resource $mimemail - A valid MIME resource.
 * @param string $data
 * @return bool - Returns TRUE on success or FALSE on failure.
 */
<<__Native>>
function mailparse_msg_parse(resource $mimemail,
                             string $data): bool;

/* Extracts/decodes a message section from the supplied filename.  The
 * contents of the section will be decoded according to their transfer
 * encoding - base64, quoted-printable and uuencoded text are supported.
 * @param resource $mimemail - A valid MIME resource, created with
 * mailparse_msg_create().
 * @param mixed $filename - Can be a file name or a valid stream resource.
 * @param mixed $callbackfunc - If set, this must be either a valid callback
 * that will be passed the extracted section, or NULL to make this function
 * return the extracted section.  If not specified, the contents will be sent
 * to "stdout".
 * @return mixed - If callbackfunc is not NULL returns TRUE on success.  If
 * callbackfunc is set to NULL, returns the extracted section as a string.
 * Returns FALSE on error.
 */
<<__Native>>
function mailparse_msg_extract_part_file(resource $mimemail,
                                         mixed $filename,
                                         mixed $callbackfunc = ""): mixed;

/* @param resource $mimemail - A valid MIME resource.
 * @param mixed $filename
 * @param mixed $callbackfunc
 * @return mixed
 */
<<__Native>>
function mailparse_msg_extract_whole_part_file(resource $mimemail,
                                               mixed $filename,
                                               mixed $callbackfunc = ""): mixed;

/* @param resource $mimemail - A valid MIME resource.
 * @param mixed $msgbody
 * @param mixed $callbackfunc
 * @return mixed - No value is returned.
 */
<<__Native>>
function mailparse_msg_extract_part(resource $mimemail,
                                    mixed $msgbody,
                                    mixed $callbackfunc = ""): mixed;

/* @param resource $mimemail
 * @return array
 */
<<__Native>>
function mailparse_msg_get_part_data(resource $mimemail): shape(
  'headers' => darray<arraykey, mixed>,
  "starting-pos" => int,
  "starting-pos-body" => int,
  "ending-pos" => int,
  "ending-pos-body" => int,
  "line-count" => int,
  "body-line-count" => int,
  "charset" => string,
  "transfer-encoding" => string,
  "content-type" => mixed,
  ?"content-disposition" => mixed,
  ?"content-location" => string,
  "content-base" => string,
  ?"content-boundary" => string,
  ?"content-id" => string,
  ?"content-description" => mixed,
  ?"content-language" => mixed,
  ?"content-md5" => mixed,
  ...
);

/* @param resource $mimemail
 * @param string $mimesection
 * @return mixed
 */
<<__Native>>
function mailparse_msg_get_part(resource $mimemail,
                                string $mimesection): mixed;

/* @param resource $mimemail
 * @return array
 */
<<__Native>>
function mailparse_msg_get_structure(resource $mimemail): darray<arraykey, string>;

/* @param string $addresses - A string containing addresses, like in: Wez
 * Furlong <wez@example.com>, doe@example.com  This string must not include
 * the header name.
 * @return array - Returns an array of associative arrays with the following
 * keys for each recipient: display The recipient name, for display purpose.
 * If this part is not set for a recipient, this key will hold the same value
 * as address. address The email address is_group TRUE if the recipient is a
 * newsgroup, FALSE otherwise.
 */
<<__Native>>
function mailparse_rfc822_parse_addresses(string $addresses): varray<shape(
  ?'display' => string,
  ?'address' => string,
  'is_group' => bool,
)>;

/* Streams data from the source file pointer, apply encoding and write to the
 * destination file pointer.
 * @param resource $sourcefp - A valid file handle. The file is streamed
 * through the parser.
 * @param resource $destfp - The destination file handle in which the encoded
 * data will be written.
 * @param string $encoding - One of the character encodings supported by the
 * mbstring module.
 * @return bool - Returns TRUE on success or FALSE on failure.
 */
<<__Native>>
function mailparse_stream_encode(resource $sourcefp,
                                 resource $destfp,
                                 string $encoding): bool;

/* Scans the data from the given file pointer and extract each embedded
 * uuencoded file into a temporary file.
 * @param resource $fp - A valid file pointer.
 * @return mixed - Returns an array of associative arrays listing filename
 * information. filename Path to the temporary file name created origfilename
 * The original filename, for uuencoded parts only The first filename entry is
 * the message body. The next entries are the decoded uuencoded files.
 */
<<__Native>>
function mailparse_uudecode_all(resource $fp): mixed;

/* Figures out the best way of encoding the content read from the given file
 * pointer.
 * @param resource $fp - A valid file pointer, which must be seek-able.
 * @return mixed - Returns one of the character encodings supported by the
 * mbstring module.
 */
<<__Native>>
function mailparse_determine_best_xfer_encoding(resource $fp): mixed;

File: hphp/runtime/ext/mailparse/ext_mailparse.cpp
/*
   +----------------------------------------------------------------------+
   | HipHop for PHP                                                       |
   +----------------------------------------------------------------------+
   | Copyright (c) 2010-present Facebook, Inc. (http://www.facebook.com)  |
   | Copyright (c) 1997-2010 The PHP Group                                |
   +----------------------------------------------------------------------+
   | This source file is subject to version 3.01 of the PHP license,      |
   | that is bundled with this package in the file LICENSE, and is        |
   | available through the world-wide-web at the following url:           |
   | http://www.php.net/license/3_01.txt                                  |
   | If you did not receive a copy of the PHP license and are unable to   |
   | obtain it through the world-wide-web, please send a note to          |
   | license@php.net so we can mail you a copy immediately.               |
   +----------------------------------------------------------------------+
*/

#include "hphp/runtime/ext/extension.h"
#include "hphp/runtime/base/temp-file.h"
#include "hphp/runtime/ext/mailparse/mime.h"
#include "hphp/runtime/ext/mailparse/rfc822.h"

namespace HPHP {

///////////////////////////////////////////////////////////////////////////////

static req::ptr<File> get_valid_file_resource(const OptResource& fp) {
  auto f = dyn_cast_or_null<File>(fp);
  if (f == nullptr || f->isClosed()) {
    raise_warning("Not a valid stream resource");
    return nullptr;
  }
  return f;
}

///////////////////////////////////////////////////////////////////////////////

OptResource HHVM_FUNCTION(mailparse_msg_create) {
  return OptResource(req::make<MimePart>());
}

bool HHVM_FUNCTION(mailparse_msg_free, const OptResource& /*mimemail*/) {
  return true;
}

Variant HHVM_FUNCTION(mailparse_msg_parse_file, const String& filename) {
  auto f = File::Open(filename, "rb");
  if (!f) return false;

  auto p = req::make<MimePart>();
  while (!f->eof()) {
    String line = f->readLine();
    if (!line.isNull()) {
      if (!MimePart::ProcessLine(p, line)) {
        return false;
      }
    }
  }
  return Variant(std::move(p));
}

bool HHVM_FUNCTION(mailparse_msg_parse,
                   const OptResource& mimemail,
                   const String& data) {
  return cast<MimePart>(mimemail)->parse(data.data(), data.size());
}

Variant HHVM_FUNCTION(mailparse_msg_extract_part_file,
                      const OptResource& mimemail,
                      const Variant& filename,
                      const Variant& callbackfunc /* = "" */) {
  return cast<MimePart>(mimemail)->
    extract(filename, callbackfunc,
            MimePart::Decode8Bit | MimePart::DecodeNoHeaders, true);
}

Variant HHVM_FUNCTION(mailparse_msg_extract_whole_part_file,
                      const OptResource& mimemail,
                      const Variant& filename,
                      const Variant& callbackfunc /* = "" */) {
  return cast<MimePart>(mimemail)->
    extract(filename, callbackfunc, MimePart::DecodeNone, true);
}

Variant HHVM_FUNCTION(mailparse_msg_extract_part,
                      const OptResource& mimemail,
                      const Variant& msgbody,
                      const Variant& callbackfunc /* = "" */) {
  return cast<MimePart>(mimemail)->
    extract(msgbody, callbackfunc,
            MimePart::Decode8Bit | MimePart::DecodeNoHeaders, false);
}

Array HHVM_FUNCTION(mailparse_msg_get_part_data, const OptResource& mimemail) {
  return cast<MimePart>(mimemail)->getPartData();
}

Variant HHVM_FUNCTION(mailparse_msg_get_part,
                      const OptResource& mimemail,
                      const String& mimesection) {
  OptResource part =
    cast<MimePart>(mimemail)->findByName(mimesection.c_str());
  if (part.isNull()) {
    raise_warning("cannot find section %s in message", mimesection.data());
    return false;
  }
  return part;
}

Array HHVM_FUNCTION(mailparse_msg_get_structure, const OptResource& mimemail) {
  return cast<MimePart>(mimemail)->getStructure();
}

const StaticString
  s_display("display"),
  s_address("address"),
  s_is_group("is_group");

Array HHVM_FUNCTION(mailparse_rfc822_parse_addresses,
                    const String& addresses) {
  php_rfc822_tokenized_t *toks =
    php_mailparse_rfc822_tokenize(addresses.data(), 1);
  php_rfc822_addresses_t *addrs = php_rfc822_parse_address_tokens(toks);

  Array ret = Array::CreateVec();
  for (int i = 0; i < addrs->naddrs; i++) {
    Array item = Array::CreateDict();
    if (addrs->addrs[i].name) {
      item.set(s_display, String(addrs->addrs[i].name, CopyString));
    }
    if (addrs->addrs[i].address) {
      item.set(s_address, String(addrs->addrs[i].address, CopyString));
    }
    item.set(s_is_group, (bool)addrs->addrs[i].is_group);
    ret.append(item);
  }

  php_rfc822_free_addresses(addrs);
  php_rfc822_tokenize_free(toks);
  return ret;
}

static int mailparse_stream_output(int c, void *stream) {
  char buf[2];
  buf[0] = c;
  buf[1] = '\0';
  return ((File*)stream)->write(buf, 1);
}
static int mailparse_stream_flush(void *stream) {
  return ((File*)stream)->flush() ? 1 : 0;
}

bool HHVM_FUNCTION(mailparse_stream_encode,
                   const OptResource& sourcefp,
                   const OptResource& destfp,
                   const String& encoding) {
  auto srcstream = get_valid_file_resource(sourcefp);
  if (!srcstream) {
    return false;
  }
  auto deststream = get_valid_file_resource(destfp);
  if (!deststream) {
    return false;
  }

  enum mbfl_no_encoding enc = mbfl_name2no_encoding(encoding.data());
  if (enc == mbfl_no_encoding_invalid)  {
    raise_warning("Unknown encoding \"%s\"", encoding.data());
    return false;
  }

  mbfl_convert_filter *conv =
    mbfl_convert_filter_new(mbfl_no_encoding_8bit, enc,
                            mailparse_stream_output, mailparse_stream_flush,
                            deststream.get());

  if (enc == mbfl_no_encoding_qprint) {
    /* If the qp encoded section is going to be digitally signed,
     * it is a good idea to make sure that lines that begin "From "
     * have the letter F encoded, so that MTAs do not stick a > character
     * in front of it and invalidate the content/signature */
    while (!srcstream->eof())  {
      String line = srcstream->readLine();
      if (!line.isNull()) {
        int i;
        if (strncmp(line.data(), "From ", 5) == 0) {
          mbfl_convert_filter_flush(conv);
          deststream->write("=46rom ", 7);
          i = 5;
        } else {
          i = 0;
        }
        const char *p = line.data();
        for (; i < line.size(); i++) {
          mbfl_convert_filter_feed(p[i], conv);
        }
      }
    }

  } else {
    while (!srcstream->eof())  {
      String data = srcstream->read();
      if (!data.empty()) {
        const char *p = data.data();
        for (int i = 0; i < data.size(); i++) {
          mbfl_convert_filter_feed(p[i], conv);
        }
      }
    }
  }

  mbfl_convert_filter_flush(conv);
  mbfl_convert_filter_delete(conv);
  return true;
}

#define UUDEC(c) (char)(((c)-' ')&077)
#define UU_NEXT(v) \
  if (line[x] == '\0' || line[x] == '\r' || line[x] == '\n') break; \
  v = line[x++]; v = UUDEC(v)

static size_t mailparse_do_uudecode(const req::ptr<File>& instream,
                                    const req::ptr<File>& outstream) {
  int A, B, C, D, n;
  size_t file_size = 0;
  if (outstream) {
    /* write to outstream */
    while (!instream->eof())  {
      String line = instream->readLine(128);
      if (line.isNull()) break;

      int x = 0;
      UU_NEXT(n);
      while (n) {
        UU_NEXT(A); UU_NEXT(B); UU_NEXT(C); UU_NEXT(D);
        if (n-- > 0) {
          file_size++;
          outstream->putc((A << 2) | (B >> 4));
        }
        if (n-- > 0) {
          file_size++;
          outstream->putc((B << 4) | (C >> 2));
        }
        if (n-- > 0) {
          file_size++;
          outstream->putc((C << 6) | D);
        }
      }
    }
  } else {
    /* skip (and measure) the data, but discard it.
     * This is separated from the version above to speed it up by a few cycles
     */
    while (!instream->eof())  {
      String line = instream->readLine(128);
      if (line.isNull()) break;

      int x = 0;
      UU_NEXT(n);
      while (line[x] && n != 0) {
        UU_NEXT(A); UU_NEXT(B); UU_NEXT(C); UU_NEXT(D);
        if (n-- > 0) file_size++;
        if (n-- > 0) file_size++;
        if (n-- > 0) file_size++;
      }
    }
  }
  return file_size;
}

const StaticString
  s_filename("filename"),
  s_origfilename("origfilename");

Variant HHVM_FUNCTION(mailparse_uudecode_all, const OptResource& fp) {
  auto instream = get_valid_file_resource(fp);
  if (!instream) {
    return false;
  }
  instream->rewind();

  auto outstream = req::make<TempFile>(false);

  Array return_value;
  int nparts = 0;
  while (!instream->eof()) {
    String line = instream->readLine();
    if (line.isNull()) break;

    /* Look for the "begin " sequence that identifies a uuencoded file */
    if (strncmp(line.data(), "begin ", 6) == 0) {

      /*
       * The next 4 bytes are an octal number for perms and a space; ignore them
       * If for some reason they aren't found treat it as an empty filename.
       */
      const char *origfilename = "";
      size_t namelen = 0;
      if (line.size() >= 10) {
        origfilename = line.data() + 10;
        namelen = line.size() - 10;
        while (namelen > 0 && isspace(origfilename[namelen-1])) {
          --namelen;
        }
      }

      /* make the return an array */
      if (nparts == 0) {
        return_value = Array::CreateVec();
        /* create an initial item representing the file with all uuencoded
           parts removed */
        Array item = Array::CreateDict();
        item.set(s_filename, String(outstream->getName()));
        return_value.append(item);
      }

      /* add an item */
      Array item = Array::CreateDict();
      item.set(s_origfilename, String(origfilename, namelen, CopyString));

      /* create a temp file for the data */
      auto partstream = req::make<TempFile>(false);
      if (partstream)  {
        nparts++;
        item.set(s_filename, String(partstream->getName()));
        return_value.append(item);

        /* decode it */
        mailparse_do_uudecode(instream, partstream);
      }
    } else {
      /* write to the output file */
      outstream->write(line);
    }
  }

  instream->rewind();
  if (nparts == 0) {
    return false;
  }
  return return_value;
}

Variant HHVM_FUNCTION(mailparse_determine_best_xfer_encoding,
                      const OptResource& fp) {
  auto stream = get_valid_file_resource(fp);
  if (!stream) {
    return false;
  }
  stream->rewind();

  int linelen = 0;
  enum mbfl_no_encoding bestenc = mbfl_no_encoding_7bit;
  bool longline = false;
  while (!stream->eof()) {
    int c = stream->getc();
    if (c > 0x80) {
      bestenc = mbfl_no_encoding_8bit;
    } else if (c == 0)  {
      bestenc = mbfl_no_encoding_base64;
      longline = false;
      break;
    }
    if (c == '\n') {
      linelen = 0;
    } else if (++linelen > 200) {
      longline = true;
    }
  }
  if (longline) bestenc = mbfl_no_encoding_qprint;
  stream->rewind();

  char * name = (char *)mbfl_no2preferred_mime_name(bestenc);
  if (name) {
    return String(name, CopyString);
  }
  return false;
}

///////////////////////////////////////////////////////////////////////////////

struct MailparseExtension final : Extension {
  MailparseExtension() : Extension("mailparse", NO_EXTENSION_VERSION_YET, NO_ONCALL_YET) { }
  void moduleInit() override {
    HHVM_FE(mailparse_msg_create);
    HHVM_FE(mailparse_msg_free);
    HHVM_FE(mailparse_msg_parse_file);
    HHVM_FE(mailparse_msg_parse);
    HHVM_FE(mailparse_msg_extract_part_file);
    HHVM_FE(mailparse_msg_extract_whole_part_file);
    HHVM_FE(mailparse_msg_extract_part);
    HHVM_FE(mailparse_msg_get_part_data);
    HHVM_FE(mailparse_msg_get_part);
    HHVM_FE(mailparse_msg_get_structure);
    HHVM_FE(mailparse_rfc822_parse_addresses);
    HHVM_FE(mailparse_stream_encode);
    HHVM_FE(mailparse_uudecode_all);
    HHVM_FE(mailparse_determine_best_xfer_encoding);
  }
} s_mailparse_extension;

///////////////////////////////////////////////////////////////////////////////
}
